lxd/storage/backend_lxd.go

```diff
@@ -3683,6 +3699,43 @@ func (b *lxdBackend) CreateInstanceSnapshot(inst instance.Instance, src instance
 		return err
 	}
 
+	// Keep track of known pools.
+	// When creating snapshots concurrently, access to this map should be locked.
+	poolMap := make(map[string]Pool)
+	poolMap[b.name] = b
+
+	// Snapshot attached volumes.
+	// TODO: Improve this by implementing concurrency.
+	for _, volume := range volumes {
+		l.Debug("Creating attached volume snapshot", logger.Ctx{"volumeProject": volume.Project, "volumePool": volume.Pool, "volumeName": volume.Name})
+
+		// Use shutdown context here as we don't have access to the request context.
+		snapshotName, err := VolumeDetermineNextSnapshotName(b.state.ShutdownCtx, b.state, volume.Pool, volume.Name, volume.Config)
+		if err != nil {
+			return err
+		}
+
+		// Add a description to easily identify attached volume snapshots.
+		description := "Attached volume snapshot related to " + inst.Type().String() + "/" + inst.Name() + " on project " + inst.Project().Name
+
+		// If pool is not known, load it.
+		volumePool, knownPool := poolMap[volume.Pool]
+		if !knownPool {
+			volumePool, err = LoadByName(b.state, volume.Pool)
+			if err != nil {
+				return err
+			}
+
+			poolMap[volume.Pool] = volumePool
+		}
+
+		// Attached volume snapshots inherit expiration time from instance snapshot.
+		err = volumePool.CreateCustomVolumeSnapshot(volume.Project, volume.Name, snapshotName, description, inst.ExpiryDate(), attachedVolumeSnapshotUUIDs[volume.Config["volatile.uuid"]], op)
```

Comment:
we should add a delete of the snapshots to the reverter so they are removed if anything goes wrong